### PR TITLE
tests: Improve schema generation test

### PR DESCRIPTION
**Pull request by nathan-stender**
Add the ability to run generate schemas script as a dry run and return the number of models changed. 

Change the test to validate that no model files are changed.

This improves test coverage and prevents the test from making changes to files.

### CODE REVIEW

File: tests/allotrope/schema_parser/generate_schemas_test.py [hunk 1]
```diff
@@ -5,4 +5,7 @@
 
 def test_generate_schemas_runs_to_completion() -> None:
     root_dir = Path(__file__).parent.parent.parent.parent
-    assert generate_schemas(root_dir) == 10
```
Comment (mdlu) [inline tests/allotrope/schema_parser/generate_schemas_test.py:8]: not having to update this int every time anymore is huge 👏

File: src/allotropy/allotrope/schema_parser/generate_schemas.py [hunk 2]
```diff
@@ -65,17 +66,17 @@ def files_equal(path1: str, path2: str) -> bool:
     return True
 
 
-def generate_schemas(root_dir: Path) -> int:
+def generate_schemas(root_dir: Path, *, dry_run: Optional[bool] = False) -> list[str]:
```
Comment (mdlu) [inline src/allotropy/allotrope/schema_parser/generate_schemas.py:69]: nit: could you add a description of what `dry_run` is? also, the name was a little confusing to me, maybe like `validate` or `validation_run` or something like that

File: tests/allotrope/schema_parser/generate_schemas_test.py [hunk 3]
```diff
@@ -5,4 +5,7 @@
...
+    models_changed = generate_schemas(root_dir, dry_run=True)
+    assert (
+        not models_changed
+    ), f"Expected no models files to have changed by generate-schemas script, found changes in: {models_changed}. Please run scripts:generate-schemas and validate the changes."
```
Comment (mdlu) [inline tests/allotrope/schema_parser/generate_schemas_test.py:11]: maybe we can have the full line to run (`hatch run scripts:generate-schemas`) in the error message so people can just copy-paste?

File: src/allotropy/allotrope/schema_parser/generate_schemas.py (hunk 2 above)
Comment (nathan-stender) [reply src/allotropy/allotrope/schema_parser/generate_schemas.py:69]: Added a description for `dry_run`. `dry_run` is pretty commonly used for scripts/commands to mean run without saving changes, so I'm inclined to keep the name.

File: tests/allotrope/schema_parser/generate_schemas_test.py (hunk 3 above)
Comment (nathan-stender) [reply tests/allotrope/schema_parser/generate_schemas_test.py]: Done!

**Review by mdlu — APPROVED** [review]

**Pull request merged**